File: vtb-project/src/runner/optimizeMADmax.cpp
#include <Eigen/Dense>
#include <chrono>
#include <iostream>
#include "params.hpp"
#include "./milp/mad_max/best_solution.hpp"
#include "./milp/mad_max/node.hpp"
#include "types.hpp"

using namespace std;
using namespace Eigen;


//for random test
int main() {
    int NN = 50;
    MatrixXd TimeSpent = MatrixXd::Zero(1, NN);
    MatrixXd Iter01proz = MatrixXd::Zero(1, NN);
    MatrixXd TimeSpent01proz = MatrixXd::Zero(1, NN);
    MatrixXd RelGap = MatrixXd::Zero(1, NN);
    MatrixXd NumIter = MatrixXd::Zero(1, NN);
    MatrixXd OptValueLow = MatrixXd::Zero(1, NN);
    MatrixXd NUMshares = MatrixXd::Zero(1, NN);
    MatrixXd TIMEhorizon = MatrixXd::Zero(1, NN);
    MatrixXd GammaMAD = MatrixXd::Zero(1, NN);

    int Nmax = 100; // maximal number of iterations

    for(int num_problem = 0; num_problem < NN; num_problem++) {
        MatrixXd historyUpperBound = MatrixXd::Zero(1, Nmax);
        MatrixXd historyLowerBound = MatrixXd::Zero(1, Nmax);
        // create parameters
        // not all parameters are used for MAD
        CommonParams* params = new CommonParams();
        params->setRandomParams();
        CommonParams_MADmax* commonParametersMax = new CommonParams_MADmax(params);
        /*std::cout << commonParametersMax->params->ind_shares << std::endl;
        std::cout << commonParametersMax->params->T << std::endl;
        std::cout << commonParametersMax->params->p_max << std::endl;
        std::cout << commonParametersMax->params->w_min << std::endl;
        std::cout << commonParametersMax->gamma_mad << std::endl;*/
        NUMshares(num_problem) = commonParametersMax->params->numShares;
        TIMEhorizon(num_problem) = commonParametersMax->params->T;
        MatrixXd historyNu = MatrixXd::Zero(commonParametersMax->params->numShares,Nmax);

        // create best integer solution
        // here we start with the empty solution with value +Inf
        auto start = std::chrono::high_resolution_clock::now();
        BestIntegerSolution* best_solution = new BestIntegerSolution(double_inf, VectorXi32::Zero(commonParametersMax->params->nu_max.size()));
        double old_best_value = best_solution->upperBound;
        //create root node
        VectorXi32 nu_lower = VectorXi32::Zero(commonParametersMax->params->numShares);
        VectorXi32 nu_upper = commonParametersMax->params->nu_max;
        Node* rootNode = new Node(commonParametersMax,nu_lower,nu_upper,best_solution,true,0);
        //starting iterations until lower bound matches upper bound
        //this happens when the root node closes
        std::clog << "Time horizon: " << commonParametersMax->params->T 
                  << ", Number of shares: " << commonParametersMax->params->numShares 
                  << ", w_min = " << commonParametersMax->params->w_min
                  << ", gammaMAD = " << commonParametersMax->gamma_mad
                  << std::endl;
        std::cout << "NumIter | Upper Bound | Lower Bound | Local Integer Solution | # integer values |  branch index | depth\n";
        int countIter = 0;
        double relGap = double_inf;
        while(rootNode->isNodeOpen && countIter < Nmax && (std::isinf(best_solution->upperBound) || relGap > 0.0001)) {
            //branch at node with minimal LP relaxation value
            Node* minNode = findMin(rootNode);
            double rV = minNode->roundValue; // when this is printed the node might no more exist
            VectorXd hN = minNode->nuRelax;
            int nDpth = minNode->depth;
            double oldval = minNode->LowerBound;
            int num_integer_values = bifurcate(minNode);
            uint bfI = minNode->BifurcationIndex;
            if(oldval > rootNode->LowerBound) {
                //std::cerr << "ERROR: oldval(" << oldval <<") > upperBound(" << rootNode->LowerBound << ")\n";
                //break;
            }

            if(old_best_value > best_solution->upperBound) {
                old_best_value = best_solution->upperBound;
                updateByUpperBound(rootNode);
            }
            countIter++;
            std::clog << countIter << " \t| "
                      << -rootNode->LowerBound << "  | "
                      << -best_solution->upperBound << "  | "
                      << -rV << "\t| "
                      << num_integer_values << "\t| "
                      << bfI << "\t| "
                      << nDpth << "\t| "
                      << std::endl;
            historyUpperBound(countIter) = -best_solution->upperBound;
            historyLowerBound(countIter) = -rootNode->LowerBound;
            historyNu.col(countIter) = hN;

            if(!std::isinf(best_solution->upperBound) && rootNode->LowerBound*best_solution->upperBound > 0) {
                relGap = rootNode->LowerBound/best_solution->upperBound - 1;
            }
            if(Iter01proz(num_problem) == 0 && relGap < 0.0001) {
                Iter01proz(num_problem) = countIter;
                
                auto finish = std::chrono::high_resolution_clock::now();
                std::chrono::duration<double> elapsed = finish - start;
                TimeSpent01proz(num_problem) = elapsed.count();
            }
        }
        if(rootNode->isNodeOpen)
            if(relGap < 0.0001 && !(best_solution->bestNu.size() == 0))
                std::clog << "Relative gap of 0.0001 achieved in " << countIter << " iterations\n";
            else 
                std::clog << "Maximal number of iterations achieved\n";
        else
            std::clog << "Optimal value: " << -best_solution->upperBound << std::endl;
        
        if (!(best_solution->bestNu.size() == 0)) {
            double MADvar = computeMADvariance(commonParametersMax, best_solution->bestNu.cast<double>());
            std::clog << "Variance of best achieved solution: " << MADvar << std::endl;
        } else {
            std::clog << "No feasible integer point found\n";
        }

        OptValueLow(num_problem) = -rootNode->LowerBound;
        GammaMAD(num_problem) = commonParametersMax->gamma_mad;

        delete commonParametersMax;
        delete best_solution;
        delete rootNode;
        auto finish = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = finish - start;
        std::clog << "Elapsed time for " << num_problem <<" problem: " << elapsed.count() << " s.\n";
        TimeSpent(num_problem) = elapsed.count();
        RelGap(num_problem) = relGap; 
        NumIter(num_problem) = countIter;
        std::cout << std::endl;
    } 
}

File: vtb-project/matlab/MADmax/optimizeMADmax.m
% main program for maximizing MAD return
close all
clear all
NN = 10;    
TimeSpent = zeros(1,NN);
Iter01proz = zeros(1,NN);
TimeSpent01proz = zeros(1,NN);
RelGap = zeros(1,NN);
NumIter = zeros(1,NN);
OptValueLow = zeros(1,NN);
NUMshares = zeros(1,NN);
TIMEhorizon = zeros(1,NN);
GammaMAD = zeros(1,NN);
for num_problem = 1:NN
fprintf("Solving problem %d\n",num_problem);
Nmax = 50; % maximal number of iterations
historyUpperBound = zeros(1,Nmax);
historyLowerBound = zeros(1,Nmax);
% create parameters
% not all parameters are used for MAD
commonParameters = create_random_instance;
if num_problem == 6
    save simpleMADmax commonParameters
elseif num_problem == 15
    save intermediateMADmax commonParameters
elseif num_problem == 71
    save difficultMADmax commonParameters
end
NUMshares(num_problem) = commonParameters.numshares;
TIMEhorizon(num_problem) = commonParameters.T;
historyNu = zeros(commonParameters.numshares,Nmax);
% create best integer solution
% here we start with the empty solution with value +Inf
tic,
[s,nu] = MADmax_feas(commonParameters);
if s
    value = commonParameters.rbar*double(nu);
    fprintf("Feasible solution has value %5e\n",value);
    best_solution = BestIntegerSolution;
    setValues(best_solution,-value,nu);
    old_best_value = best_solution.upperBound;
    % create root node
    nu_lower = uint16(zeros(commonParameters.numshares,1));
    nu_upper = commonParameters.numax;
    rootNode = NodeMADmax(commonParameters,nu_lower,nu_upper,best_solution,true,0);
    % starting iterations
    % until lower bound matches upper bound
    % this happens when the root node closes
    fprintf("Starting main optimization\n");
    fprintf("NumIter | Upper Bound  | Lower Bound  | Local Integer Solution | # l_i = u_i | # u_i - l_i = 1 | # integer values | branch index | depth\n");
    countIter = 0;
    relGap = +Inf;
    while rootNode.nodeOpen && (countIter < Nmax) && (relGap > 0.0001)
        % branch at node with minimal LP relaxation value
        minNode = findMin(rootNode);
        rV = minNode.roundValue; % when this is printed the node might no more exist
        hN = minNode.nuRelax;
        nDpth = minNode.depth;
        n0 = length(minNode.determined_nu);
        n1 = length(find(minNode.LowerNuBound + 1 == minNode.UpperNuBound));
        oldval = rootNode.LowerBound;
        oldTab = minNode.simplexTableau;
        oldBas = minNode.basicSet;
        [num_integer_values,bfI,bfV] = bifurcate(minNode);
%         if isempty(bfI)
%             disp('error')
%         end
%         if oldval > rootNode.LowerBound
%             disp('error')
%         end
        if old_best_value > best_solution.upperBound
            old_best_value = best_solution.upperBound;
            updateByUpperBound(rootNode);
        end
        countIter = countIter + 1;
        fprintf("  %5d | %5e | %5e |           %5e |        %5d    |        %5d    |        %5d     |     %5d    | %5d\n",countIter,-rootNode.LowerBound,-best_solution.upperBound,-rV,n0,n1,num_integer_values,bfI,nDpth);
        historyUpperBound(countIter) = -best_solution.upperBound;
        historyLowerBound(countIter) = -rootNode.LowerBound;
        historyNu(:,countIter) = hN;
        if rootNode.LowerBound*best_solution.upperBound > 0
            relGap = rootNode.LowerBound/best_solution.upperBound - 1;
        end
        if (Iter01proz(num_problem) == 0) && (relGap < 0.001)
            Iter01proz(num_problem) = countIter;
            TimeSpent01proz(num_problem) = toc;
        end
    end
    % clear objects
    if rootNode.nodeOpen
        if (relGap < 0.0001) && (~isempty(best_solution.bestNu))
            fprintf("Relative gap of 0.0001 achieved in %d iterations\n",countIter);
        else
            fprintf("Maximal number of iterations achieved\n");
        end
    else
        fprintf("Optimal value: %5e\n",-best_solution.upperBound);
    end
    if ~isempty(best_solution.bestNu)
        MADvar = computeMADvariance(commonParameters,best_solution.bestNu);
        fprintf("Variance of best achieved solution: %5e\n",MADvar);
    end
end
if ~s || isempty(best_solution.bestNu)
    fprintf("No feasible integer point found")
end
if s
    OptValueLow(num_problem) = -rootNode.LowerBound;
end
GammaMAD(num_problem) = commonParameters.gammaMAD;
delete(rootNode);
delete(best_solution);
delete(commonParameters);
timeSpent = toc;
fprintf("Time spent: %5e sec\n\n",timeSpent);
TimeSpent(num_problem) = timeSpent;
RelGap(num_problem) = relGap;
NumIter(num_problem) = countIter;
end
k = NN;
csvwrite('MADmax_tmp.csv',[1:k; TIMEhorizon(1:k); NUMshares(1:k); TimeSpent01proz(1:k); TimeSpent(1:k); RelGap(1:k); Iter01proz(1:k); NumIter(1:k); GammaMAD(1:k)]');

% figure
% plot((1:countIter)/countIter*timeSpent,historyUpperBound(1:countIter))
% hold on
% plot((1:countIter)/countIter*timeSpent,historyLowerBound(1:countIter))
% xlabel('time (sec)')
% title('Upper and lower bounds')
% grid on
% figure
% plot((1:countIter)/countIter*timeSpent,log(max(zeros(1,countIter),historyLowerBound(1:countIter)./historyUpperBound(1:countIter) - 1))/log(10))
% title('Relative Log Gap (Basis 10)')
% xlabel('time (sec)')
% grid on
